source/opt/simplification_pass.cpp

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "simplification_pass.h"
+
+#include <set>
+#include <unordered_set>
+#include <vector>
+
+#include "fold.h"
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status SimplificationPass::Process(ir::IRContext* c) {
+  InitializeProcessing(c);
+  bool modified = false;
+
+  for (ir::Function& function : *get_module()) {
+    modified |= SimplifyFunction(&function);
+  }
+  return (modified ? Status::SuccessWithChange : Status::SuccessWithoutChange);
+}
+
+bool SimplificationPass::SimplifyFunction(ir::Function* function) {
+  bool modified = false;
+  // Phase 1: Traverse all instructions in dominance order.
+  // The second phase will only be on the instructions whose inputs have changed
+  // after being processed during phase 1.  Since OpPhi instructions are the
+  // only instructions whose inputs do not necessarily dominate the use, we keep
+  // track of the OpPhi instructions already seen, and add them to the work list
+  // for phase 2 when needed.
+  std::vector<ir::Instruction*> work_list;
+  std::unordered_set<ir::Instruction*> process_phis;
+  std::unordered_set<ir::Instruction*> inst_to_kill;
+  std::unordered_set<ir::Instruction*> in_work_list;
+
+  cfg()->ForEachBlockInReversePostOrder(
+      function->entry().get(),
+      [&modified, &process_phis, &work_list, &in_work_list, &inst_to_kill,
+       this](ir::BasicBlock* bb) {
+        for (ir::Instruction* inst = &*bb->begin(); inst;
+             inst = inst->NextNode()) {
+          if (inst->opcode() == SpvOpPhi) {
+            process_phis.insert(inst);
+          }
+
+          if (inst->opcode() == SpvOpCopyObject || FoldInstruction(inst)) {
+            modified = true;
+            context()->AnalyzeUses(inst);
+            get_def_use_mgr()->ForEachUser(
+                inst, [&work_list, &process_phis,
+                       &in_work_list](ir::Instruction* use) {
+                  if (process_phis.count(use) && !in_work_list.count(use)) {
```

Comment:
... && !in_work_list.insert(use).second))  {

Allows you to remove the subsequent insert.

Reply:
fixed.